stability: protect against recreation of GCed txn records

tbg opened the pull request:
Fix the main issue in #9265. When a transaction is aborted before
having written its initial transaction record, intents written by the
transaction before creating its record may be aborted.

If the GC queue deletes that ABORTED entry, there must be a mechanism
which prevents the original transaction to create what it considers the
"initial" transaction record, or that transaction could commit successfully
but lose some its writes.

Such a mechanism is added here: The GC queue keeps time of the threshold
used for grooming the transaction span and communicates this into the
replicated Range state, where BeginTransaction can check it and return
an appropriate error.

Fixes #9265.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9377)

<!-- Reviewable:end -->

spencerkimball commented on keys/constants.go:
```diff
@@ -106,6 +106,9 @@ var (
 	LocalLeaseAppliedIndexSuffix = []byte("rlla")
 	// localRangeStatsSuffix is the suffix for range statistics.
 	LocalRangeStatsSuffix = []byte("stat")
+	// LocalTxnSpanThresholdSuffix is the suffix for the last txn span GC's
```
From the comment, this should probably be `LocalTxnSpanGCThresholdSuffix`. Doesn't feel like there's otherwise enough context in the name.

spencerkimball commented on keys/keys.go:
```diff
@@ -198,6 +198,12 @@ func RangeLastGCKey(rangeID roachpb.RangeID) roachpb.Key {
 
+// RangeTxnSpanThresholdKey returns a system-local key for last used GC
+// threshold on the txn span.
+func RangeTxnSpanThresholdKey(rangeID roachpb.RangeID) roachpb.Key {
```
`TxnSpanGC`

spencerkimball commented on roachpb/api.go:
```diff
@@ -770,7 +770,10 @@ func (*AdminTransferLeaseRequest) flags() int { return isAdmin | isAlone }
-func (*RangeLookupRequest) flags() int        { return isRead | isTxn }
+
+// TODO(tschottdorf): why does RangeLookupRequest have the isTxn flag?
+func (*RangeLookupRequest) flags() int { return isRead | isTxn }
```
The `isTxn` flag makes no sense here and should be removed. I tracked down the one usage in `DistSender` and it completely orthogonal to `RangeLookupRequest`. Remove the flag.

tschottdorf commented:
With hindsight, the main issue might well be not setting `OrigTimestamp` when creating the aborted entry, but that too is fixed (and tested) here.

spencerkimball commented on storage/gc_queue.go:
```diff
@@ -559,8 +563,8 @@ func RunGC(
-	gcKeys = append(gcKeys, processAbortCache(ctx, snap, desc.RangeID, now,
-		abortCacheAgeThreshold, &infoMu, pushTxnFn)...)
+	gcKeys = append(gcKeys, processAbortCache(
+		ctx, snap, desc.RangeID, abortSpanThreshold, &infoMu, pushTxnFn)...)
```
Might be nice to have 'GC' in the name: `abortSpanGCThreshold`. It just feel ambiguous without.

spencerkimball commented on storage/intent_resolver.go:
```diff
@@ -289,11 +289,17 @@ func (ir *intentResolver) processIntentsAsync(r *Replica, intents []intentsWithA
-				// transaction already having been aborted by someone else, in
-				// which case the client may still be running. Thus, we must
-				// poison.
+				// we don't guarantee that). But there's others paths in which
```
s/there's others/there are other/

spencerkimball commented on storage/intent_resolver.go:
```diff
@@ -289,11 +289,17 @@ func (ir *intentResolver) processIntentsAsync(r *Replica, intents []intentsWithA
-				// poison.
+				// we don't guarantee that). But there's others paths in which
+				// the Push comes back successful while the coordinating client
+				// may still be active. Examples of this are when
```
s/when/when:/

spencerkimball commented on storage/replica_command.go:
```diff
@@ -467,6 +467,20 @@ func (r *Replica) BeginTransaction(
+	threshold := r.mu.state.TxnSpanThreshold
+	r.mu.Unlock()
+
+	// Disallow a transaction record to be created if it's at a timestamp below
```
How about "Disallow creation of transaction record if it's at a timestamp before..."

spencerkimball commented on storage/replica_command.go:
```diff
@@ -1297,6 +1325,8 @@ func (r *Replica) PushTxn(
 		// If getting an update for a transaction record which doesn't yet
 		// exist, return empty Pushee.
+		//
+		// TODO(tschottdorf): Explain why PUSH_TOUCH is not special-cased.
```
Not sure what you mean by "special-cased"?

spencerkimball commented:
Still haven't looked at unittest in detail but out of time.

spencerkimball commented:
Unittest LGTM

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[storage/gc_queue.go, line 355 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi6o7lradqlH8DT_ga:-KRi6o7lradqlH8DT_gb:byrfuit) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/gc_queue.go#L355)):*
> ```Go
> 	// potentially necessary intent resolutions did not fail).
> 	TransactionSpanGCAborted, TransactionSpanGCCommitted, TransactionSpanGCPending int
> 	// TxnSpanThreshold is the timestamp below the last activity of which
> ```

this sentence is confused. how about

TxnSpanThreshold is the high water mark below which transaction span entries have been considered for GC.

---

*[storage/replica_test.go, line 2443 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi7btjWdHHqUczLlR-:-KRi7btjWdHHqUczLlR0:b6zvhhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2443)):*
> ```Go
> // verify that if the GC queue could potentially have removed a txn record
> // created through a successful push (by a concurrent actor), the original
> // transaction's subsequent attempts to create its initial record fail.
> ```

s/fail/fails/

---

*[storage/replica_test.go, line 2454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi7ejyRmBflbipyUdH:-KRi7ejyRmBflbipyUdI:bym7x3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2454)):*
> ```Go
> 	key := roachpb.Key("a")
> 	desc := tc.rng.Desc()
> 	tc.manualClock.Increment(123)
> ```

why? comment.

---

*[storage/replica_test.go, line 2468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi7nXRozEzoBtpHpvP:-KRi7nXRozEzoBtpHpvQ:b-230c8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2468)):*
> ```Go
> 	abortedPushee := resp.(*roachpb.PushTxnResponse).PusheeTxn
> 	if abortedPushee.Status != roachpb.ABORTED {
> 		t.Fatalf("expected push to abort pushee, got %+v", resp)
> ```

nit: s/resp/abortedPushee/

---

*[storage/replica_test.go, line 2505 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi86zFQaE_x0ecFy1Y:-KRi86zFQaE_x0ecFy1Z:b-4xu2sb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2505)):*
> ```Go
> 		beginArgs, header := beginTxnArgs(key, pushee)
> 		resp, pErr := tc.SendWrappedWith(header, &beginArgs)
> 		if _, ok := pErr.GetDetail().(*roachpb.TransactionAbortedError); !ok {
> ```

if `pErr` is `nil` here, you'll get an NPE on `GetDetail()`, which is probably not what you want.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-:-KRi8EoR0CAIXU9qVNq_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: protect against recreation of GCed txn records

Fix the main issue in #9265. When a transaction is aborted before
having written its initial transaction record, intents written by the
transaction before creating its record may be aborted.

If the GC queue deletes that ABORTED entry, there must be a mechanism
which prevents the original transaction to create what it considers the
"initial" transaction record, or that transaction could commit successfully
but lose some its writes.

Such a mechanism is added here: The GC queue keeps time of the threshold
used for grooming the transaction span and communicates this into the
replicated Range state, where BeginTransaction can check it and return
an appropriate error.

Fixes #9265.

tschottdorf commented:
Review status: 3 of 23 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[keys/constants.go, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7Ig7-r1-109:-KRi3HeNYqr7aWsgdW4z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/keys/constants.go#L109)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> From the comment, this should probably be `LocalTxnSpanGCThresholdSuffix`. Doesn't feel like there's otherwise enough context in the name.
</details>
Done.

---

*[keys/keys.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7Ig8-r1-203:-KRi3I87zx3DxK72HPdw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/keys/keys.go#L203)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> `TxnSpanGC`
</details>
Done.

---

*[roachpb/api.go, line 775 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7IgB-r1-775:-KRiFCANukTTd22-6Er5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/roachpb/api.go#L775)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> The `isTxn` flag makes no sense here and should be removed. I tracked down the one usage in `DistSender` and it completely orthogonal to `RangeLookupRequest`. Remove the flag.
</details>
Done.

---

*[storage/gc_queue.go, line 355 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi6o7lradqlH8DT_ga:-KRiB7XvZ96OHI69t9Vg:bfgvc6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/gc_queue.go#L355)):*
That sentence loses the vital information (what of the txn do we compare against this threshold?). Changed t

> 	// TxnSpanGCThreshold is the cutoff for transaction span GC. Transactions
	// with a smaller LastActive() are considered for GC.

---

*[storage/gc_queue.go, line 567 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7IgF-r1-567:-KRiBpvNMS8lPGCQ36_3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/gc_queue.go#L567)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Might be nice to have 'GC' in the name: `abortSpanGCThreshold`. It just feel ambiguous without.
</details>
Done.

---

*[storage/intent_resolver.go, line 292 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7IgH-r1-292:-KRiC-gCOb6S7V9MNtlq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/intent_resolver.go#L292)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/there's others/there are other/
</details>
Done.

---

*[storage/intent_resolver.go, line 294 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7IgH-r1-294:-KRiC1AtmU1slSkCYaoO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/intent_resolver.go#L294)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/when/when:/
</details>
Done.

---

*[storage/replica_command.go, line 474 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7IgI-r1-474:-KRiCAiW7jq_-aEHEgi3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_command.go#L474)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> How about "Disallow creation of transaction record if it's at a timestamp before..."
</details>
Done.

---

*[storage/replica_command.go, line 1329 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KReWyKE7PgcyOBJ7IgI-r1-1329:-KRiCGBEWAoMcEvjcBnI:b-sulscs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_command.go#L1329)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Not sure what you mean by "special-cased"?
</details>
Expanded (and TODO removed)

---

*[storage/replica_test.go, line 2443 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi7btjWdHHqUczLlR-:-KRiD4RznIVs1s6lGlL-:bjz12we) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2443)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/fail/fails/
</details>
The sentence was ok as is, but gave it a touch up.

---

*[storage/replica_test.go, line 2454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi7ejyRmBflbipyUdH:-KRiDcXdffKTM9V_cj3Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2454)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why? comment.
</details>
Done.

---

*[storage/replica_test.go, line 2468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi7nXRozEzoBtpHpvP:-KRiDGNMTsLGW14NwMdz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2468)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: s/resp/abortedPushee/
</details>
Done.

---

*[storage/replica_test.go, line 2505 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi86zFQaE_x0ecFy1Y:-KRiE9mio5hepUPRs4WT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/replica_test.go#L2505)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> if `pErr` is `nil` here, you'll get an NPE on `GetDetail()`, which is probably not what you want.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->

tbg committed:
roachpb: remove isTxn flag from RangeLookup

This flag is only queried in `IsPossibleTransaction()` in the
DistSender (in the context of disallowing range-spanning ops
without a txn), and does not apply to RangeLookup which we
never send transactionally (and which the executing code does
not expect to have a transaction set).

tamird commented:
Reviewed 19 of 19 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[roachpb/api.proto, line 438 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiMZQLwqkKpR0AZpfs:-KRiMZQNXxKND1_aZ4vJ:bqwkdk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/roachpb/api.proto#L438)):*
> ```protobuf
>   // considered for GC (and thus might have been removed).
>   optional util.hlc.Timestamp txn_span_gc_threshold = 5 [(gogoproto.nullable) = false,
>                                                          (gogoproto.customname) = "TxnSpanGCThreshold" ];
> ```

nonstandard indent here

---

*[storage/gc_queue.go, line 355 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi6o7lradqlH8DT_ga:-KRiMeUIIP9YhZxXToGp:b2s4djn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/gc_queue.go#L355)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> That sentence loses the vital information (what of the txn do we compare against this threshold?). Changed t
> 
> > 	// TxnSpanGCThreshold is the cutoff for transaction span GC. Transactions
> 	// with a smaller LastActive() are considered for GC.
</details>
"are considered" is still not right. either "were considered" or "will be considered"

also, please re-wrap this.

---

*[storage/storagebase/state.proto, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiNA025H1-w-dV8jqH:-KRiNA025H1-w-dV8jqI:b1x4f15) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/storage/storagebase/state.proto#L59)):*
> ```protobuf
>   // transaction entry.
>   util.hlc.Timestamp txn_span_gc_threshold = 9 [(gogoproto.nullable) = false,
>                                                 (gogoproto.customname) = "TxnSpanGCThreshold"];
> ```

same nonstandard indent here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

tschottdorf commented:
Review status: 13 of 23 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 438 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiMZQLwqkKpR0AZpfs:-KRicv1Hl6ExgBDqUwbs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/roachpb/api.proto#L438)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nonstandard indent here
</details>
Done.

---

*[storage/gc_queue.go, line 355 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRi6o7lradqlH8DT_ga:-KRid6GIYg-tUw52HYY4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e51513d6caf8aac3a92c5313d519e82b03d96f9c/storage/gc_queue.go#L355)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "are considered" is still not right. either "were considered" or "will be considered"
> 
> also, please re-wrap this.
</details>
Done.

---

*[storage/storagebase/state.proto, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiNA025H1-w-dV8jqH:-KRifY0bhMs8bzNszJ00:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/storage/storagebase/state.proto#L59)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> same nonstandard indent here
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 10 of 10 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 438 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiMZQLwqkKpR0AZpfs:-KRihrxuvBQSUt2I-K4v:bulygo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/roachpb/api.proto#L438)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done.
</details>
Is it?

---

*[storage/storagebase/state.proto, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiNA025H1-w-dV8jqH:-KRihzPvRDl12yXCMVDs:bulygo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/storage/storagebase/state.proto#L59)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done.
</details>
Is it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 438 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiMZQLwqkKpR0AZpfs:-KRii9YlEisKhJgFWwnG:byaa3jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/roachpb/api.proto#L438)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Is it?
</details>
also trailing space before the closing bracket

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 438 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiMZQLwqkKpR0AZpfs:-KRij27TeWaAX_4FyteR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/roachpb/api.proto#L438)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> also trailing space before the closing bracket
</details>
Done.

---

*[storage/storagebase/state.proto, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377#-KRiNA025H1-w-dV8jqH:-KRiik7GU1uEwRbL-Da3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b5dbd399d3c746869d71e20fe5563d8b2ecb93/storage/storagebase/state.proto#L59)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Is it?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 4 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9377)*
<!-- Sent from Reviewable.io -->
